### PR TITLE
Use zsh 5.4's WARN_NESTED_VAR, convert driver to anonymous function

### PR DESCRIPTION
**Pull request by danielshahaf**
If this is merged before upstream 5.4 is released then update #412 at merge time.

### CODE REVIEW

**Comment by nicoulaj** [comment]
(Closing/reopening to trigger continuous integration)

File: zsh-syntax-highlighting.zsh [hunk 1]
```diff
@@ -28,13 +29,17 @@
 # -------------------------------------------------------------------------------------------------
 
 # First of all, ensure predictable parsing.
-zsh_highlight__aliases=`builtin alias -Lm '[^+]*'`
+local zsh_highlight__aliases="`builtin alias -Lm '[^+]*'`"
```
Comment (phy1729) [inline zsh-syntax-highlighting.zsh]: Since you're touching this line anyway?, can you convert backticks to `$( )`?

File: zsh-syntax-highlighting.zsh [hunk 2]
```diff
@@ -28,13 +29,17 @@
...
 # In zsh <= 5.2, `alias -L` emits aliases that begin with a plus sign ('alias -- +foo=42')
 # them without a '--' guard, so they don't round trip.
 #
 # Hence, we exclude them from unaliasing:
 builtin unalias -m '[^+]*'
 
+# Set the 5.4-to-be WARN_NESTED_VAR option if it's available.
+emulate -L zsh
+setopt localoptions warncreateglobal ${(k)options[(I)warnnestedvar]}
```
Comment (phy1729) [inline zsh-syntax-highlighting.zsh]: `$options` requires zsh/parameter. Perhaps you could use `zsyh_user_options`?

File: zsh-syntax-highlighting.zsh (hunk 1 above)
Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: Touché.

Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: Err, UI bug. I meant to say: Sure can; what's the rationale?

File: zsh-syntax-highlighting.zsh (hunk 2 above)
Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: Touché.

File: zsh-syntax-highlighting.zsh (hunk 1 above)
Comment (phy1729) [reply zsh-syntax-highlighting.zsh]: Consistency with the rest of the code and `$( )` nests better. I generally replace any backticks I come across.

Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: I'm not as zealous, but sure, I'll have it use the house style.

Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: Done.

File: zsh-syntax-highlighting.zsh [hunk 3]
```diff
@@ -1,3 +1,4 @@
+() { # that's here to make the line numbers in errors match the file line numbers.
```
Comment (phy1729) [inline zsh-syntax-highlighting.zsh]: This doesn't appear to be true. If I checkout `ca6b31f`, when I open a new terminal I get
```
(anon):46: no such file or directory: /home/phy1729/.version
(anon):47: no such file or directory: /home/phy1729/.revision-hash
zsh-syntax-highlighting: highlighters directory '/home/phy1729/highlighters' not found.
```
 however lines 46-48 are
```
  # $0 is reliable
  typeset -g ZSH_HIGHLIGHT_VERSION=$(<"${0:A:h}"/.version)
  typeset -g ZSH_HIGHLIGHT_REVISION=$(<"${0:A:h}"/.revision-hash)
```
The line numbers seem to be off by one.

File: zsh-syntax-highlighting.zsh [hunk 4]
```diff
@@ -45,7 +50,7 @@ if true; then
     # When running from a source tree without 'make install', $ZSH_HIGHLIGHT_REVISION
     # would be set to '$Format:%H$' literally.  That's an invalid value, and obtaining
     # the valid value (via `git rev-parse HEAD`, as Makefile does) might be costly, so:
-    ZSH_HIGHLIGHT_REVISION=HEAD
+    typeset -g ZSH_HIGHLIGHT_REVISION=HEAD
```
Comment (phy1729) [inline zsh-syntax-highlighting.zsh]: Is this change needed? It's `typeset -g` a few lines above.

File: zsh-syntax-highlighting.zsh [hunk 5]
```diff
@@ -397,11 +402,14 @@ zmodload zsh/parameter 2>/dev/null || true
 autoload -U is-at-least
 
 # Initialize the array of active highlighters if needed.
-[[ $#ZSH_HIGHLIGHT_HIGHLIGHTERS -eq 0 ]] && ZSH_HIGHLIGHT_HIGHLIGHTERS=(main)
+[[ $#ZSH_HIGHLIGHT_HIGHLIGHTERS -eq 0 ]] && ZSH_HIGHLIGHT_HIGHLIGHTERS[1]=main
```
Comment (phy1729) [inline zsh-syntax-highlighting.zsh]: Curious what's the motivation for this change.

**Comment by danielshahaf** [comment]
Discussed on IRC.  This is too untidy and will be done as part of #252.

File: zsh-syntax-highlighting.zsh (hunk 4 above)
Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: Good catch.

File: zsh-syntax-highlighting.zsh (hunk 5 above)
Comment (danielshahaf) [reply zsh-syntax-highlighting.zsh]: No idea!
